File: cli/bench/judge.py
#!/usr/bin/env python3
"""Grade the non-exact hits of a recorded bench run, evals spec section 4.4.

One run of `grammachy bench --record <dir>` writes `<dir>/checks.json`, one
entry per engine, model, and item, carrying the item and the answer. This
script reads that file, keeps every non-exact hit, folds the identical answers
of two models onto one item, and asks Claude Fable 5 one question per folded
item. It writes `judgements.json` beside the record, keyed by item id and then
result text, which is what `grammachy bench --judgements <file>` reads.

A non-exact hit is a valid Check on an item that carries a mistake, where at
least one Issue touches a span the item expects and applying every Fix does not
reproduce the expected sentence. An item nothing touched is a plain miss: the
writer is offered nothing to accept, so there is nothing to grade.

The call is lean, which is the build item HUF-210 recorded. A full Claude Code
session cost about 0.25 USD of notional spend per item on the pilot, which does
not scale to the 365-item eval set. So every call disables the tools, the MCP
servers, the skills, and the project settings, and replaces the default system
prompt with one sentence.

A judgements file costs real spend, so a later run adds to the file rather than
replacing it.
This run folds its own answers over the ones already there, by item id and
result text.
The write lands on a `.pending` sibling and is renamed, the rule `checks.json`
follows.
So a run that dies part way leaves the earlier file whole.
A run that graded nothing refuses to touch a file that exists.
Pass `--replace` to overwrite the file wholesale instead.

Usage:

    cli/bench/judge.py <dir-or-checks.json> [--out FILE] [--jobs N]
                       [--model NAME] [--limit N] [--dry-run] [--replace]

`--dry-run` prints the folded items and the first prompt and calls nothing, so
the selection can be read without spending anything.
"""

from __future__ import annotations

import argparse
import concurrent.futures
import json
import os
import subprocess
import sys
from pathlib import Path

# The judge of spec section 4.4. Fable 5 is what HUF-210 measured against the
# hand labels, so a different model invalidates the 88% the gate was set from.
MODEL = "claude-fable-5"

# One sentence in place of the whole Claude Code system prompt.
SYSTEM_PROMPT = (
    "You grade one grammar-checker suggestion for usefulness to the writer. "
    "Answer with one JSON object and nothing else."
)

# The question HUF-210 settled, meaning clause included: the judge's stricter
# calls on the pilot were a lost object and a missing preposition, both real
# faults, so "correct English, or clearly better and not broken" stays.
QUESTION = (
    "Question: would a writer be helped by accepting these edits? "
    "Useful means the result is correct English, or clearly better than the "
    "original and not broken. Not useful means the result is wrong, broken, or "
    "worse than the original. The result need not equal the reference.\n"
    'Answer with only a JSON object: {"useful": true or false, "reason": "one sentence"}'
)

# How long one call may take before it is counted as a failure.
CALL_TIMEOUT_S = 300


def overlaps(a_start: int, a_end: int, b_start: int, b_end: int) -> bool:
    """Whether two half-open spans share at least one code unit."""
    return a_start < b_end and b_start < a_end


def is_caught(check: dict) -> bool:
    """Whether at least one Issue touches a span the item expects."""
    return any(
        overlaps(issue["start"], issue["end"], edit["start"], edit["end"])
        for issue in check["issues"]
        for edit in check["edits"]
    )


def collapse(text: str) -> str:
    """Runs of whitespace collapsed, the comparison of spec section 3."""
    return " ".join(text.split())


def non_exact_hits(checks: list[dict]) -> list[dict]:
    """Every Check of the record that is a non-exact hit under the default.

    A record entry that carries a `thinking` field is kept only when it ran
    under the product default: thinking on for a local row, and whatever the
    provider pins for a cloud row. A record written before that field existed
    ran under the default of its own run, so it is kept as it stands.
    """
    kept = []
    for check in checks:
        if not check.get("valid") or not check.get("edits"):
            continue
        if check.get("result_text") is None:
            continue
        if check.get("thinking") is False and check["engine"] != "openrouter":
            continue
        if not is_caught(check):
            continue
        if collapse(check["result_text"]) == collapse(check["expected_text"]):
            continue
        kept.append(check)
    return kept


def fold(hits: list[dict]) -> list[dict]:
    """One item per (item id, result text), with the rows that produced it.

    Two models that answer one sentence the same way are one judgement and one
    call, which is what keeps the eval set affordable.
    """
    folded: dict[tuple[str, str], dict] = {}
    for hit in hits:
        key = (hit["id"], hit["result_text"])
        item = folded.get(key)
        if item is None:
            item = {
                "id": hit["id"],
                "native": hit["native"],
                "text": hit["text"],
                "expected_text": hit["expected_text"],
                "edits": hit["edits"],
                "result_text": hit["result_text"],
                "issues": hit["issues"],
                "rows": [],
            }
            folded[key] = item
        row = f"{hit['engine']}:{hit['model']}"
        if row not in item["rows"]:
            item["rows"].append(row)
    return list(folded.values())


def prompt_for(item: dict) -> str:
    """The one prompt of spec section 4.4, in the wording HUF-210 measured."""
    suggested = [
        {
            "original": issue["original"],
            "fix": issue["fix"],
            "reason": issue["reason"],
        }
        for issue in item["issues"]
    ]
    reference = [{"text": edit["text"], "fix": edit["fix"]} for edit in item["edits"]]
    return (
        f"The writer's native language: {item['native']}.\n"
        f"Original sentence: {json.dumps(item['text'])}\n"
        f"Reference correction: {json.dumps(item['expected_text'])}\n"
        f"The edits the reference makes: {json.dumps(reference)}\n"
        f"The checker's suggested edits: {json.dumps(suggested)}\n"
        "The sentence the writer gets after accepting every edit: "
        f"{json.dumps(item['result_text'])}\n\n"
        f"{QUESTION}"
    )


def command(model: str, prompt: str) -> list[str]:
    """The lean `claude -p` invocation of spec section 4.4.

    No tools, no MCP server, no skill, no project or local settings file, and
    one sentence of system prompt. The session is not persisted, because a
    judgement run of the eval set would otherwise leave hundreds of them.
    """
    return [
        "claude",
        "-p",
        "--model",
        model,
        "--output-format",
        "json",
        "--system-prompt",
        SYSTEM_PROMPT,
        "--tools",
        "",
        "--strict-mcp-config",
        "--mcp-config",
        '{"mcpServers": {}}',
        "--setting-sources",
        "",
        "--disable-slash-commands",
        "--no-session-persistence",
        prompt,
    ]


def read_answer(stdout: str) -> tuple[bool, str, float | None]:
    """The judgement inside one `--output-format json` answer.

    `--output-format json` prints either one result object or the whole event
    array, depending on the version, and a wrapper such as `mise` may print a
    line of its own first. So the parse starts at the first bracket of either
    kind rather than at the first character. The model's own text is the JSON
    object the prompt asked for, found by its braces, because a model may still
    wrap its object in a sentence.

    `useful` must be a JSON boolean. A model asked for `true or false` may write
    the string `"false"`, which is truthy, so a coercion would record the
    opposite label and hand it to the gate with no error at all.
    """
    starts = [at for at in (stdout.find("["), stdout.find("{")) if at >= 0]
    if not starts:
        raise ValueError(f"no JSON in the answer: {stdout[:200]!r}")
    envelope = json.loads(stdout[min(starts) :])
    if isinstance(envelope, list):
        envelope = [entry for entry in envelope if entry.get("type") == "result"][-1]
    text = envelope.get("result", "")
    answer = json.loads(text[text.index("{") : text.rindex("}") + 1])
    useful = answer["useful"]
    if not isinstance(useful, bool):
        raise ValueError(f"useful is not a boolean: {useful!r}")
    return useful, str(answer.get("reason", "")), envelope.get("total_cost_usd")


def call_context(finished: subprocess.CompletedProcess | None) -> str:
    """What a failed `claude` call left behind, for the summary `main` prints.

    A bad model name, an expired login, and a rate limit all reach the parser as
    an empty stdout, so the exit code and the tail of stderr are the only way to
    tell them apart without running the whole grading again.
    """
    if finished is None:
        return ""
    return f" (exit {finished.returncode}, stderr: {finished.stderr.strip()[:200]!r})"


def judge(item: dict, model: str) -> dict:
    """Grade one folded item, or carry the reason it could not be graded."""
    finished = None
    try:
        finished = subprocess.run(
            command(model, prompt_for(item)),
            capture_output=True,
            text=True,
            timeout=CALL_TIMEOUT_S,
        )
        useful, reason, cost = read_answer(finished.stdout)
        return {"item": item, "useful": useful, "reason": reason, "cost_usd": cost}
    except Exception as failure:  # noqa: BLE001 - one bad call must not end the run
        error = f"{type(failure).__name__}: {failure}"[:300] + call_context(finished)
        return {"item": item, "error": error}


def read_judgements(out: Path) -> dict[str, dict[str, dict]]:
    """The judgements already in the file, or nothing when it is not one.

    A truncated write, an empty file, or a hand edit all leave something this
    script cannot merge. That is a file with nothing to keep rather than a
    reason to end the run, so the operator is told and the run writes over it.
    """
    try:
        existing = json.loads(out.read_text())
    except (OSError, ValueError) as failure:
        print(f"{out} cannot be read, so it is replaced: {failure}", file=sys.stderr)
        return {}
    if not isinstance(existing, dict) or not all(
        isinstance(answers, dict) for answers in existing.values()
    ):
        print(f"{out} is not a judgements file, so it is replaced.", file=sys.stderr)
        return {}
    return existing


def write_judgements(out: Path, judgements: dict, replace: bool) -> int:
    """Fold this run's judgements over the file already there, and say how many.

    A judgement was paid for, so a run that graded a different slice of the
    record adds to the file rather than emptying it. `--replace` is the one way
    to drop what is there, for the run that means to.

    The write lands on a `.pending` sibling and is renamed, so a run that dies
    part way leaves the earlier file whole.

    A file this script cannot read is merged as if it were absent. There is
    nothing to keep from it, and refusing would throw away the calls this run
    just paid for, so it says so on stderr and writes anyway.
    """
    merged: dict[str, dict[str, dict]] = {}
    if out.exists() and not replace:
        merged = read_judgements(out)
    kept = sum(len(answers) for answers in merged.values())
    for item_id, answers in judgements.items():
        merged.setdefault(item_id, {}).update(answers)

    pending = out.with_name(out.name + ".pending")
    pending.write_text(json.dumps(merged, indent=2, ensure_ascii=False) + "\n")
    os.replace(pending, out)
    return kept


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "record",
        type=Path,
        help="the --record directory of a bench run, or its checks.json",
    )
    parser.add_argument(
        "--out",
        type=Path,
        help="where to write judgements.json (default: beside checks.json)",
    )
    parser.add_argument("--model", default=MODEL, help=f"the judge (default {MODEL})")
    parser.add_argument("--jobs", type=int, default=4, help="calls in flight at once")
    parser.add_argument("--limit", type=int, help="grade only the first N items")
    parser.add_argument(
        "--dry-run",
        action="store_true",
        help="print the selection and the first prompt, and call nothing",
    )
    parser.add_argument(
        "--replace",
        action="store_true",
        help="overwrite the output file rather than adding this run to it",
    )
    arguments = parser.parse_args()

    checks_path = arguments.record
    if checks_path.is_dir():
        checks_path = checks_path / "checks.json"
    checks = json.loads(checks_path.read_text())

    items = fold(non_exact_hits(checks))
    items.sort(key=lambda item: (item["id"], item["result_text"]))
    if arguments.limit is not None:
        items = items[: arguments.limit]
    print(
        f"{len(checks)} Checks, {len(items)} folded non-exact hits to judge.",
        file=sys.stderr,
    )

    if arguments.dry_run:
        for item in items:
            print(f"{item['id']}\t{item['rows']}\t{item['result_text']}")
        if items:
            print("\n--- the first prompt ---\n" + prompt_for(items[0]))
        return 0

    with concurrent.futures.ThreadPoolExecutor(arguments.jobs) as pool:
        graded = list(pool.map(lambda item: judge(item, arguments.model), items))

    judgements: dict[str, dict[str, dict]] = {}
    failures = []
    spend = 0.0
    for answer in graded:
        item = answer["item"]
        if "error" in answer:
            failures.append(f"{item['id']}: {answer['error']}")
            continue
        judgements.setdefault(item["id"], {})[item["result_text"]] = {
            "useful": answer["useful"],
            "reason": answer["reason"],
        }
        spend += answer["cost_usd"] or 0.0

    out = arguments.out or checks_path.with_name("judgements.json")
    refused = not judgements and out.exists() and not arguments.replace
    if refused:
        print(
            f"Nothing was graded, so {out} keeps every judgement it already holds.\n"
            "Pass --replace to empty it on purpose.",
            file=sys.stderr,
        )
    else:
        kept = write_judgements(out, judgements, arguments.replace)
        print(f"{out} now holds this run's judgements over {kept} earlier ones.", file=sys.stderr)
    print(
        f"{len(items) - len(failures)} judged, {len(failures)} failed, "
        f"{spend:.2f} USD notional.",
        file=sys.stderr,
    )
    for failure in failures:
        print(f"  {failure}", file=sys.stderr)
    return 1 if failures or refused else 0


if __name__ == "__main__":
    sys.exit(main())

File: cli/src/bench/judge.rs
//! The Useful fix column and its agreement gate, evals spec section 4.4.
//!
//! A model that finds the right span but writes a different correction scores
//! zero on exact fix, and the pilot found that a good part of those answers
//! still help the writer. So `cli/bench/judge.py` grades every non-exact hit of
//! a recorded run with Claude Fable 5 and writes `judgements.json`. This module
//! is the reading half: it turns that file into the Useful fix column and
//! decides whether the column may count in the ranking.
//!
//! - A **non-exact hit** is an interference item whose valid Check touched at
//!   least one expected span but whose Fixes do not reproduce `expected_text`.
//!   An item nothing touched is a plain miss and is never judged, because the
//!   writer is offered nothing to accept.
//! - Both files are keyed by **item id, then result text**, the sentence the
//!   writer gets after Accept. Nesting the two keys is what keeps a result text
//!   that holds any delimiter readable, and it is what folds the identical
//!   answers of two models onto one judgement.
//! - The **gate** is spec section 4.4: the column counts in the ranking only
//!   when the judge agrees with the committed hand labels on at least 80% of
//!   the labelled items of that run, over a sample of at least
//!   [`MINIMUM_LABELLED`] of them. Below the gate, under that sample, or with
//!   no label matched, the column still prints and the file says it is
//!   excluded.
//!
//! The hand labels are compiled in from `tests/fixtures/judge-labels.json`, so
//! a released binary needs no repository checkout, the same rule the fixture
//! follows. They carry no eval-set text of their own: every label of the file
//! is a fixture item, and section 2.1 forbids committing FCE text.

use std::collections::BTreeMap;

use serde::Deserialize;

/// The agreement a run needs before the Useful fix column may rank, in percent.
pub const AGREEMENT_GATE: f64 = 80.0;

/// The smallest labelled sample the agreement gate may be measured on.
///
/// A result text has to match a hand label verbatim, so a run matches only a
/// few of the 17 committed labels. One matched label at 100% agreement is not
/// evidence that the judge is sound, and the ranking swap it would unlock
/// decides the recommendation, so a run under this sample keeps the raw
/// ranking.
pub const MINIMUM_LABELLED: usize = 5;

/// The committed hand labels, the truth the judge is measured against.
const LABELS: &str = include_str!(concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/tests/fixtures/judge-labels.json"
));

/// One graded answer: was the writer helped by accepting these edits.
#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
pub struct Judgement {
    pub useful: bool,
    /// One sentence from the judge. A hand label carries none.
    #[serde(default)]
    pub reason: String,
}

/// A judgements or hand-label file: item id, then result text, then the answer.
pub type Judgements = BTreeMap<String, BTreeMap<String, Judgement>>;

/// Read a judgements file, or say why it is not one.
pub fn read(path: &std::path::Path) -> Result<Judgements, String> {
    let text = std::fs::read_to_string(path)
        .map_err(|error| format!("--judgements: {} cannot be read: {error}", path.display()))?;
    parse(&text).map_err(|error| {
        format!(
            "--judgements: {} is not a judgements file: {error}",
            path.display()
        )
    })
}

/// Read one judgements file from its text.
pub fn parse(text: &str) -> Result<Judgements, serde_json::Error> {
    serde_json::from_str(text)
}

/// The committed hand labels of spec section 4.4.
pub fn labels() -> Judgements {
    parse(LABELS).expect("the compiled hand labels are a judgements file")
}

/// One non-exact hit of a run, folded onto its judgement key.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Hit {
    /// The engine and model whose row produced it, the key of the table row.
    pub row: (String, String),
    pub id: String,
    /// The sentence the writer gets after Accept.
    pub result: String,
}

/// The Useful fix count of one Models row.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub struct RowTally {
    /// Non-exact hits the judgements file graded useful.
    pub useful: usize,
    /// Non-exact hits the judgements file graded at all.
    pub judged: usize,
    /// Non-exact hits the row produced, graded or not.
    pub hits: usize,
}

impl RowTally {
    /// The Useful fix cell, `n of m (rate)` over the judged hits.
    ///
    /// A row whose hits are all missing from the file prints what is missing
    /// rather than a rate of nothing, because a silent `0 of 0` reads as a
    /// model that helped nobody.
    pub fn cell(&self) -> String {
        if self.judged == 0 {
            return match self.hits {
                0 => "no non-exact hit".to_string(),
                hits => format!("not judged ({hits} hits)"),
            };
        }
        format!(
            "{} of {} ({:.1}%)",
            self.useful,
            self.judged,
            100.0 * self.useful as f64 / self.judged as f64
        )
    }
}

/// What one judgements file says about one run.
#[derive(Debug, Clone, Default)]
pub struct Assessment {
    /// The Useful fix count of every row that produced a non-exact hit.
    rows: BTreeMap<(String, String), RowTally>,
    /// Folded non-exact hits of the run that a hand label also covers.
    pub labelled: usize,
    /// Labelled hits where the judge and the hand label say the same thing.
    pub agreed: usize,
    /// Folded non-exact hits of the run, and how many the file graded.
    pub hits: usize,
    pub judged: usize,
}

impl Assessment {
    /// Grade one run: the hits it produced against the judgements it was given.
    ///
    /// The row counts are per row, because the column is a row column. The
    /// agreement is over the folded hits of the whole run, because the gate is
    /// a fact about the judge on this run rather than about one model.
    pub fn of(hits: &[Hit], judgements: &Judgements, labels: &Judgements) -> Assessment {
        let mut assessment = Assessment::default();
        let mut folded: BTreeMap<(&str, &str), ()> = BTreeMap::new();

        for hit in hits {
            let judgement = look_up(judgements, &hit.id, &hit.result);
            let row = assessment.rows.entry(hit.row.clone()).or_default();
            row.hits += 1;
            if let Some(judgement) = judgement {
                row.judged += 1;
                if judgement.useful {
                    row.useful += 1;
                }
            }

            // The agreement counts one folded hit once, however many rows
            // produced the same answer for the same item.
            if folded
                .insert((hit.id.as_str(), hit.result.as_str()), ())
                .is_some()
            {
                continue;
            }
            assessment.hits += 1;
            let Some(judgement) = judgement else { continue };
            assessment.judged += 1;
            let Some(label) = look_up(labels, &hit.id, &hit.result) else {
                continue;
            };
            assessment.labelled += 1;
            if judgement.useful == label.useful {
                assessment.agreed += 1;
            }
        }

        assessment
    }

    pub fn row(&self, engine: &str, model: &str) -> Option<RowTally> {
        self.rows
            .get(&(engine.to_string(), model.to_string()))
            .copied()
    }

    /// How often the judge and the hand labels said the same thing, in percent.
    ///
    /// A run no hand label covers has no agreement to report, which is not the
    /// same as an agreement of zero.
    pub fn agreement_percent(&self) -> Option<f64> {
        if self.labelled == 0 {
            return None;
        }
        Some(100.0 * self.agreed as f64 / self.labelled as f64)
    }

    /// Whether the Useful fix column may count in the ranking.
    ///
    /// The gate is both a rate and a sample: a judge measured on fewer than
    /// [`MINIMUM_LABELLED`] labels is unproven whatever its agreement.
    pub fn ranks(&self) -> bool {
        self.labelled >= MINIMUM_LABELLED
            && self
                .agreement_percent()
                .is_some_and(|percent| percent >= AGREEMENT_GATE)
    }

    /// The sentences the benchmark file carries under the Quality table.
    pub fn lines(&self) -> String {
        let mut out = String::new();
        out.push_str(&format!(
            "Useful fix: {} of the {} folded non-exact hits of this run carry a judgement in the judgements file.\n",
            self.judged, self.hits
        ));
        match self.agreement_percent() {
            Some(_) if self.labelled < MINIMUM_LABELLED => out.push_str(&format!(
                "This run matched {}, under the {MINIMUM_LABELLED} the {AGREEMENT_GATE:.0}% gate needs, so the judge is unproven here.\nThe report keeps the raw ranking of exact fix rate.\n",
                match self.labelled {
                    1 => "1 hand label".to_string(),
                    labelled => format!("{labelled} hand labels"),
                }
            )),
            Some(percent) if self.ranks() => out.push_str(&format!(
                "The judge agreed with the hand labels on {} of {} ({percent:.1}%), at or above the {AGREEMENT_GATE:.0}% gate, so the Useful fix column counts in the ranking.\n",
                self.agreed, self.labelled
            )),
            Some(percent) => out.push_str(&format!(
                "The judge agreed with the hand labels on {} of {} ({percent:.1}%), under the {AGREEMENT_GATE:.0}% gate, so the Useful fix column is printed but excluded from the ranking.\n",
                self.agreed, self.labelled
            )),
            None => out.push_str(&format!(
                "No hand label of `cli/tests/fixtures/judge-labels.json` covers a hit of this run, so the {AGREEMENT_GATE:.0}% gate could not be measured and the Useful fix column is excluded from the ranking.\n",
            )),
        }
        out
    }
}

/// The judgement of one item and result text, when the file carries it.
fn look_up<'a>(judgements: &'a Judgements, id: &str, result: &str) -> Option<&'a Judgement> {
    judgements.get(id)?.get(result)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn hit(engine: &str, model: &str, id: &str, result: &str) -> Hit {
        Hit {
            row: (engine.to_string(), model.to_string()),
            id: id.to_string(),
            result: result.to_string(),
        }
    }

    /// A judgements file in the shape `judge.py` writes.
    fn file(entries: &[(&str, &str, bool)]) -> Judgements {
        let mut out = Judgements::new();
        for (id, result, useful) in entries {
            out.entry(id.to_string()).or_default().insert(
                result.to_string(),
                Judgement {
                    useful: *useful,
                    reason: "a recorded reason".to_string(),
                },
            );
        }
        out
    }

    #[test]
    fn a_row_counts_the_useful_share_of_the_hits_that_were_judged() {
        let hits = [
            hit("openai", "gemma", "zh-03", "kept one"),
            hit("openai", "gemma", "zh-04", "kept two"),
            hit("openai", "gemma", "zh-06", "never judged"),
        ];
        let judgements = file(&[("zh-03", "kept one", true), ("zh-04", "kept two", false)]);

        let row = Assessment::of(&hits, &judgements, &Judgements::new())
            .row("openai", "gemma")
            .expect("the row produced hits");

        assert_eq!(
            row,
            RowTally {
                useful: 1,
                judged: 2,
                hits: 3
            }
        );
        assert_eq!(row.cell(), "1 of 2 (50.0%)");
    }

    #[test]
    fn a_row_with_no_judged_hit_says_so_rather_than_printing_a_rate() {
        assert_eq!(
            RowTally {
                useful: 0,
                judged: 0,
                hits: 2
            }
            .cell(),
            "not judged (2 hits)"
        );
        assert_eq!(RowTally::default().cell(), "no non-exact hit");
    }

    #[test]
    fn agreement_counts_one_folded_hit_once_however_many_rows_produced_it() {
        let hits = [
            hit("openai", "gemma", "zh-03", "same answer"),
            hit("openrouter", "deepseek", "zh-03", "same answer"),
        ];
        let judgements = file(&[("zh-03", "same answer", true)]);
        let labels = file(&[("zh-03", "same answer", true)]);

        let assessment = Assessment::of(&hits, &judgements, &labels);

        assert_eq!((assessment.hits, assessment.judged), (1, 1));
        assert_eq!((assessment.agreed, assessment.labelled), (1, 1));
        // Both rows still carry the hit, because the column is a row column.
        assert_eq!(assessment.row("openai", "gemma").unwrap().hits, 1);
        assert_eq!(assessment.row("openrouter", "deepseek").unwrap().hits, 1);
    }

    /// A run of `labels` hand labels where the first `agree` of them agree.
    fn run_of(labels: usize, agree: usize) -> Assessment {
        let entries: Vec<(String, String)> = (0..labels)
            .map(|n| (format!("zh-0{n}"), format!("answer {n}")))
            .collect();
        let judged: Vec<(&str, &str, bool)> = entries
            .iter()
            .map(|(id, result)| (id.as_str(), result.as_str(), true))
            .collect();
        let hits: Vec<Hit> = judged
            .iter()
            .map(|(id, result, _)| hit("openai", "gemma", id, result))
            .collect();
        // A label the judge disagrees with is the flipped answer.
        let labelled: Vec<(&str, &str, bool)> = judged
            .iter()
            .enumerate()
            .map(|(index, (id, result, _))| (*id, *result, index < agree))
            .collect();

        Assessment::of(&hits, &file(&judged), &file(&labelled))
    }

    #[test]
    fn a_judge_that_agrees_with_four_of_five_labels_clears_the_gate() {
        let assessment = run_of(5, 4);

        assert_eq!((assessment.agreed, assessment.labelled), (4, 5));
        assert_eq!(assessment.agreement_percent(), Some(80.0));
        assert!(assessment.ranks(), "80% is at the gate, not under it");
    }

    #[test]
    fn a_judge_that_agrees_with_three_of_five_labels_is_under_the_gate() {
        let assessment = run_of(5, 3);

        assert_eq!(assessment.agreement_percent(), Some(60.0));
        assert!(!assessment.ranks());
        assert!(
            assessment.lines().contains("under the 80% gate"),
            "{}",
            assessment.lines()
        );
    }

    /// A gate measured on a handful of labels proves nothing, so it never
    /// unlocks the ranking swap however well the judge did on them.
    #[test]
    fn a_judge_that_matched_four_labels_is_unproven_however_well_it_agreed() {
        let assessment = run_of(4, 4);

        assert_eq!(assessment.agreement_percent(), Some(100.0));
        assert!(!assessment.ranks(), "four labels is under the sample");
        assert!(
            assessment.lines().contains(
                "This run matched 4 hand labels, under the 5 the 80% gate needs, so the judge is unproven here.\nThe report keeps the raw ranking of exact fix rate.\n"
            ),
            "{}",
            assessment.lines()
        );
    }

    #[test]
    fn five_labels_are_the_smallest_sample_the_gate_is_measured_on() {
        assert!(run_of(5, 5).ranks());
        assert!(!run_of(4, 4).ranks());
    }

    #[test]
    fn a_run_no_hand_label_covers_reports_no_agreement_and_never_ranks() {
        let hits = [hit("openai", "gemma", "zh-03", "an answer")];
        let judgements = file(&[("zh-03", "an answer", true)]);

        let assessment = Assessment::of(&hits, &judgements, &Judgements::new());

        assert_eq!(assessment.agreement_percent(), None);
        assert!(!assessment.ranks());
        assert!(
            assessment.lines().contains("could not be measured"),
            "{}",
            assessment.lines()
        );
    }

    /// A hand label keyed by the same item but a different result text is a
    /// different answer, so it must not be read as this one's truth.
    #[test]
    fn a_label_on_another_result_of_the_same_item_is_not_this_hit_s_label() {
        let hits = [hit("openai", "gemma", "zh-03", "what this model wrote")];
        let judgements = file(&[("zh-03", "what this model wrote", true)]);
        let labels = file(&[("zh-03", "what another model wrote", false)]);

        let assessment = Assessment::of(&hits, &judgements, &labels);

        assert_eq!(assessment.labelled, 0);
    }

    #[test]
    fn the_committed_hand_labels_are_the_seventeen_pilot_items() {
        let labels = labels();
        let items: usize = labels.values().map(BTreeMap::len).sum();
        let useful: usize = labels
            .values()
            .flat_map(BTreeMap::values)
            .filter(|label| label.useful)
            .count();

        assert_eq!(items, 17, "the pilot sample of HUF-210");
        assert_eq!(useful, 8, "eight useful and nine not useful");
    }

    /// The committed labels carry no eval-set text, the rule of section 2.1.
    ///
    /// Every label of the file belongs to an item of the fixture, which is
    /// committed text already. A label on a fetched eval-set item would put
    /// FCE-derived text in the repository, so it never belongs here.
    #[test]
    fn every_committed_hand_label_names_a_fixture_item() {
        let ids: Vec<String> = crate::bench::fixture::sentences()
            .into_iter()
            .map(|sentence| sentence.id)
            .collect();

        for id in labels().keys() {
            assert!(ids.contains(id), "{id} is not a fixture item");
        }
    }
}
